Allow copying paths from the File Tab context menu

HarshNarayanJha opened the pull request:
Pretty bare-bones implementation of copying paths from tab context menu. ~Right now, both actions do the same thing (i.e. copy relative path), as I need to find a way to get the absolute path~ Figured out!

Release Notes:

- Fixed ... ([#13970](https://github.com/zed-industries/zed/issues/13970)).

HarshNarayanJha committed:
fix: desktop file TryExec missing

HarshNarayanJha committed:
Merge branch 'zed-industries:main' into main

HarshNarayanJha committed:
feat: implement copy path and copy path relative for file tabs (aka pane)

NOTE: copy path is not absolute right now, need to find a way to copy absolute paths

HarshNarayanJha committed:
cargo fmt

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -152,6 +152,8 @@ actions!(
         SplitRight,
         SplitDown,
         TogglePreviewTab,
+        CopyPath,
```
We already have `outline_panel::CopyRelativePath` and `project_panel::CopyRelativePath`, and this will add another relatively specific namespace, `pane::`, into the mix.

I would rather use a generic `workspace::` one for these new two (same as https://github.com/zed-industries/zed/pull/13976/files#diff-041673bbd1947a35d45945636c0055429dfc8b5985faf93f8a8a960c9ad31e28 does) 

So that later we could change both panels to use that `workspace::CopyPath` and `worksace::CopyRelativePath` pair, and future places like editor can reuse the same `workspace::` action too.

HarshNarayanJha replied:
Okay, I will add those two to the workspace crate

HarshNarayanJha replied:
Done!

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -1119,6 +1121,28 @@ impl Pane {
+        if let Some(clipboard_text) = self
+            .active_item()
+            .as_ref()
+            .and_then(|entry| entry.project_path(cx))
```
Both code places use the same `project_path` which is wrong.
What is meant to happen is an absolute path retrieval? You have to do `worktree.absolutize` (there's one place in the pane.rs that does this).

HarshNarayanJha replied:
Right I said that
> Right now, both actions do the same thing (i.e. copy relative path), as I need to find a way to get the absolute path
I now I know how to do that

HarshNarayanJha replied:
Now it copies absolute path!

williamdes commented:
Can you also post a screenshot please?

HarshNarayanJha commented:
Sure!

![20240726_16h01m19s_grim](https://github.com/user-attachments/assets/d3d881b9-3c94-41c5-8ac8-c81bedc992e7)

<details>
<summary>Read only if you want to answer</summary>

Btw, why is the target folder 16GB after compiling? And even if I change a line in `pane.rs`, upon running again using cargo run, why it again compiles some of the crates which I haven't touched? (It always starts compiling from workspace crate, 1145/1193, and then takes a lot of time and CPU, and heats up my laptop).
Any way to change this behaviour?

I know rust projects are heavy, but this is too much for my laptop. I only started learning rust a few months ago and didn't expect this much from big projects 😂. Likewise, I plan to move zed repo to my external ssd to save space.

Another thing, I am using stable zed to edit the files, which is a little unusable since cargo checks _every_ crate upon a single edit, which slows down the zed, and my laptop. I need to disable checking while writing code. _Thank god I didn't open it up in VSC_

</details>

williamdes commented:
>Btw, why is the target folder 16GB after compiling? And even if I change a line in pane.rs, upon running again using cargo run, why it again compiles some of the crates which I haven't touched? (It always starts compiling from workspace crate, 1145/1193, and then takes a lot of time and CPU, and heats up my laptop).
Any way to change this behaviour?

Yeah, quite a pain
`cargo clean` is your friend 
@SomeoneToIgnore do you know why Zed re builds all crates each time? 

>I know rust projects are heavy, but this is too much for my laptop. I only started learning rust a few months ago and didn't expect this much from big projects 😂. Likewise, I plan to move zed repo to my external ssd to save space.

haha, same

>Another thing, I am using stable zed to edit the files, which is a little unusable since cargo checks every crate upon a single edit, which slows down the zed, and my laptop. I need to disable checking while writing code. Thank god I didn't open it up in VSC

Hmm, I think I have found an option in the settings.json for the rust language that disables this.
Having tons of projects in my forked folder it tries to compile all of them. This is a pain.
Will send you the option here when I am back

SomeoneToIgnore commented:
<details>
  <summary>These questions indeed belong to some #new-rustaceans generic channel, but let's do that once</summary>
  
  > Btw, why is the target folder 16GB after compiling?

This means you've compiled it too few times, you need to try a few more times, with different toolchain versions preferrably
```
du -h target/
...snip
189G	target
```

cargo and rustc need various intermediate data for incremental compilation and build creation.
In addition to that, default debug builds contain a lot of "extra" things like debug information, not-fully-optimized-and-reduced code (that could cargo --release or a differently configured linker do at cost of much more CPU usage), and Zed uses a lot of dependencies, exposing such issues more than inside an average rust project.

Whenever something in the build chain diverges (a crate version, a rustc version, etc.) — whatever's affected by that change has to be recompiled, bringing in more artefacts, debug info, incremental info, etc.
As cargo does not know the semantics of such changes, it preserves old versions (what if you just switch a branch and now switching back, wanting faster recompilation?) of old files, sometimes too aggressively by my taste.

`cargo clean` or, better, https://github.com/holmgr/cargo-sweep might help here

> And even if I change a line in `pane.rs`, upon running again using cargo run, why it again compiles some of the crates which I haven't touched? 

pane.rs is a part of a quite foundational crate, `workspace`, and now every crate that depends on `workspace`, transitively or not, has to recheck whether it used (transitively or not) certain modules that you've changed or not — and if you did, it has to recompile.
And, of course, the final binary has to be relinked nonetheless.

> and then takes a lot of time and CPU, and heats up my laptop).
Any way to change this behaviour?

Realistically — using a different machine, either a remote one (Zed has some remote capabilities enough to work and it improves every month).

Apart from that, you could install a different linker if you're on Linux (Zed CI uses `mold`) and tinker with cargo profiles more, adding another near https://github.com/zed-industries/zed/blob/20213a7ed0c4d3c6bf2d1f8179d617ce2b32e058/Cargo.toml#L506 with something, that improves your situation (no good suggestions, docs and cargo --timings to profile your local build).

After that — only some work on Zed's dependencies to reduce them, but not sure what are bottlenecks on your machine.

> which is a little unusable since cargo checks every crate upon a single edit

There are two things: your default settings save files very often, presumably after editor's focus is lost (which might be good by itself) — that triggers the default feature of rust-analyzer, "cargo check on save", which produces most of its diagnostics currently.

So, either you could move into state of manual saves only, or tweak check on save (up to disabling it fully): https://zed.dev/docs/languages/rust#large-projects-and-performance
  
</details>

HarshNarayanJha committed:
refactor: add CopyPath and CopyRelativePath to workspace::

HarshNarayanJha committed:
now copy path copies absolute path instead of relative path thanks to `worktree.absolutize()`

add a new `paths` variable to include both relative and absolute paths in `pane.rs`, and use those paths in the context menu option.

HarshNarayanJha commented:
Thanks very much for the help, Someone! rust-analzer options have been very helpful, I can now use intellisense (though a bit outdated), but it works!

Now it's ready to merge. I may also create the PR to unify `Copy{Path,RelativePath}` into `workspace::`, if possible.

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -1119,6 +1119,19 @@ impl Pane {
         )
     }
 
+    fn copy_path(&mut self, _: &CopyPath, path: PathBuf, cx: &mut ViewContext<Self>) {
```
These two methods are useless now, they do the exact same oneliner thing with `cx.write_to_clipboard(ClipboardItem::new(path.to_string_lossy().to_string()));`.

They accept different arguments for some reason, but that changes nothing.
Let's either remove both methods entirely or get the abs path inside one of them.

HarshNarayanJha replied:
I was trying to get absolute path out of the entry and didn't notice they did the same thing… I will move that line to the call site and remove the functions then

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -1761,30 +1774,64 @@ impl Pane {
+                                    }),
+                                )
+                            })
+                            .when_some(rel_path, |menu, path| {
```
`rel_path` always exists, why do we need to wrap it with some `when_some`?

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -1761,30 +1774,64 @@ impl Pane {
+                                    cx.handler_for(&pane, move |pane, cx| {
+                                        pane.copy_relative_path(
+                                            &CopyRelativePath,
+                                            path.to_path_buf(),
```
> to_path_buf

This is an extra allocation for nothing, as later all we do with it is `.to_string_lossy().to_string()`.
A `&Path` or a `&PathBuf` is more than just enough.

HarshNarayanJha replied:
Removing both methods and calling clipboard inline resolves this

SomeoneToIgnore commented on crates/workspace/src/pane.rs:
```diff
@@ -1761,30 +1774,64 @@ impl Pane {
                             .ok()
                             .flatten();
 
+                        let (rel_path, abs_path) =
```
All these lines below is a very dirty code, I liked the previous version more.

Why do we have to drag the `entry.path` out of all this while `if let Some(entry) = single_entry_to_resolve` is the source of the very same relative path?

Cannot we just leave the `let parent_abs_path = ...` exactly as it was before and get the relative path out of the entry?

HarshNarayanJha replied:
This is partially because I couldn't figure out how to get `entry.path` out of it, so I tried to use tuple as the return type, and later deconstruct it, and create new `parent_abs_path` for the `Open In Terminal` to use.

Though even if we revert those lines, I can think of no way to `absolute_path` and `relative_path` out of the clouser. And the `entry` is not the same since it's being redeclared by `entry = worktree.entry_for_id(entry)?;`
so in order to get the absolute path, I have to repeat those lines.

I think the better solution is to return a tuple of relative_path, absolute_path, and parent_abs_path from the clouser and use them

```rust
let (abs_path, rel_path, parent_abs_path) = pane
    .update(cx, |pane, cx| {
        pane.workspace.update(cx, |workspace, cx| {
            let project = workspace.project().read(cx);
            project.worktree_for_entry(entry, cx).and_then(|worktree| {
                let worktree = worktree.read(cx);
                let entry = worktree.entry_for_id(entry)?;
                let abs_path = worktree.absolutize(&entry.path).ok()?;

                let abs_path = Some(abs_path.to_path_buf());
                let rel_path = entry.path.to_path_buf();

                let parent = if entry.is_symlink {
                    abs_path.canonicalize().ok()?
                } else {
                    abs_path
                }
                .parent()?
                .to_path_buf();

                Some((abs_path, rel_path, parent))
            })
        })
    })
    .ok()
    .flatten()
    .map(|(abs_path, rel_path, parent)| (abs_path, rel_path, parent));
```

SomeoneToIgnore replied:
I think we can leave that block alone, it is quite packed already.

Why cannot you extract a common method for retrieving an absolute path for a given entry? Then you could do `.and_then(|abs_path| abs_path|.parent())` on it in the original block, and use it right away in the new function.

SomeoneToIgnore replied:
Or just repeat the abs path extraction in your separate function for handling the "copy path" action?

HarshNarayanJha replied:
I was afraid of repeating a pre-existing block into my own function, but if I can extract and use, then...

<details>
<summary>For readability</summary>

```rust
...
let absolute_path = Pane::get_absolute_path(&pane, cx, entry); 
let parent_abs_path = Pane::get_absolute_path(&pane, cx, entry)
.and_then(|abs_path| Some(abs_path.parent()?.to_path_buf()));
...
```
In place of that `quite packed` block inside `render_tab`

The new (or old) get_absolute_path function
```rust
fn get_absolute_path(
        pane: &View<Pane>,
        cx: &mut WindowContext,
        entry: ProjectEntryId,
    ) -> Option<PathBuf> {
        pane.update(cx, |pane, cx| {
            pane.workspace.update(cx, |workspace, cx| {
                let project = workspace.project().read(cx);
                project.worktree_for_entry(entry, cx).and_then(|worktree| {
                    let worktree = worktree.read(cx);
                    let entry = worktree.entry_for_id(entry)?;
                    let abs_path = worktree.absolutize(&entry.path).ok()?;
                    let path = if entry.is_symlink {
                        abs_path.canonicalize().ok()?
                    } else {
                        abs_path
                    };

                    Some(path)
                })
            })
        })
        .ok()
        .flatten()
    }
```

Old `copy_relative_path`
```rust
fn copy_relative_path(&mut self, _: &CopyRelativePath, cx: &mut ViewContext<Self>) {
        if let Some(clipboard_text) = self
            .active_item()
            .as_ref()
            .and_then(|entry| entry.project_path(cx))
            .map(|p| p.path.to_string_lossy().to_string())
        {
            cx.write_to_clipboard(ClipboardItem::new(clipboard_text));
        }
    }
```

And finally the menu entries themselves,
```rust
...
menu = menu
    .separator()
    .when_some(absolute_path, |menu, path| {
        menu.entry(
            "Copy Path",
            Some(Box::new(CopyPath)),
            cx.handler_for(&pane, move |_, cx| {
                cx.write_to_clipboard(ClipboardItem::new(
                    path.to_string_lossy().to_string(),
                ));
            }),
        )
    })
    .entry(
        "Copy Relative Path",
        Some(Box::new(CopyRelativePath)),
        cx.handler_for(&pane, move |pane, cx| {
            pane.copy_relative_path(&CopyRelativePath, cx);
        }),
    )
    ...
```

</details>

HarshNarayanJha replied:
I hope this time it is all right, I have had compilation errors quite a few times now. Learnt a lot of rust by refactoring it out 😓.
Sorry for this many re-reviews

SomeoneToIgnore reviewed (commented):
I really liked the version from https://github.com/HarshNarayanJha/zed/blob/212605984629f8c9cc7019577bd14f497e4a85d1/crates/workspace/src/pane.rs more, now it's a bunch of questionable code added which is harder to read.

HarshNarayanJha committed:
refactor: create fn Pane::get_absolute_path and copy_relative_path

also re-create copy_relative_path into it's own function, while the menu entry itself handles copying absolute path (by the value from Pane::get_absolute_path)

HarshNarayanJha commented:
Hope this helps!

SomeoneToIgnore committed:
Style fixes

SomeoneToIgnore committed:
More style fixes

SomeoneToIgnore reviewed (approved):
Thank you.
I've pushed some style fixes as it seemed faster to do it myself than commenting.

Now it looks fine to merge.

HarshNarayanJha commented:
Why is this PR not in the latest release?

SomeoneToIgnore commented:
Hello again.
https://github.com/zed-industries/zed/issues/15148#issuecomment-2262212225
